Chore(app): media server base image

tomaszdurka opened the pull request.

tomaszdurka committed:
Initial commit with base image for karma-media-server

tomaszdurka committed:
Introduce production target for image

roosnic1 commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  "main": "./dist/",
+  "license": "UNLICENSED",
+  "scripts": {
+    "start": "node ./dist/index.js",
```
I'd add here a setup script also to copy `.env.example` to `.env` for dockerless development.

roosnic1 commented on .env.example:
```diff
@@ -0,0 +1,4 @@
+TOKEN=secret
```
I think it's cleaner if we use the same `TOKEN` throughout the whole repository.

roosnic1 commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+FROM node:12.20.1-alpine AS app
```
i suggest to use the latest `node:12-alpine` base image

roosnic1 commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+FROM app as app-production
+
+ENV NODE_ENV=production
+RUN yarn install
```
is this going to remove the `devDependencies` which were installed on line 37?

tomaszdurka replied:
Yep.

roosnic1 commented on README.md:
```diff
@@ -1 +1,29 @@
+# karma-media-server
+
+## Development
+
```
I suggest to add a short section for local development also:

```
- yarn install
- yarn setup
- yarn build
- yarn setup
```

roosnic1 commented on README.md:
````diff
@@ -1 +1,29 @@
+
+### Setup
+```
+docker-compose run --rm media yarn install
````
why the run command? Doesn't that happen on `docker-compose up`?

tomaszdurka replied:
Problem is that when running `docker-compose up`:
- Image is built, including `yarn install` and is fully ready;
- But when you run local image with volume attached the `node_modules` is taken from host machine and is empty;
- Then I run `yarn install` in docker so it installs it via volume on host machine - that's only for dev so we have local `node_modules` present
- I don't run `build` as it's started with `watch` anyway.

roosnic1 replied:
👍🏾

roosnic1 replied:
@tomaszdurka had the following error on my machine:

`error An unexpected error occurred: "EACCES: permission denied, mkdir '/home/node/node_modules/abbrev'".` 

is that my docker setup or can we prevent this with some parameters in the volume mounts?

tomaszdurka replied:
Aw, seems this kind of problems are not present on mac hosts. Seems you need provide user and group for that docker runs. As I cannot reproduce this right now and be sure, can you try adding the following `docker-compose.override.yml` file and running `up` again?
```
services:
  media:
    user: "<your-id>:<your-group>"
```

Use values from running:
```
$ id
```

roosnic1 commented on README.md:
````diff
@@ -1 +1,29 @@
+
+### Manual upload test
+```
+curl $(docker-compose port media 4001) -X POST -H "Authorization: Bearer super-secret" -F file=@package.json
````
I think it would be nice if we had here also the expected result of the curl. And does the media server accept *.json files? 😄

tomaszdurka replied:
It actually does support other files :) 

I've tested that with a GET request right after using `jq`, but didn't want to add `jq` to extract id easily, so could add something like this:
```
HOST=$(docker-compose port media 4001)
ID=$(curl $HOST -X POST -H "Authorization: Bearer secret" -F file=@package.json | jq -r .id)
curl $HOST/$ID
```

Let me know if you like that direction.

roosnic1 replied:
@tomaszdurka  I actually mean more to include the response of the `curl` call in the `README.md`. Something linke this:

If everything was successful you should see a response like:
```
{"id":"QtAIafvvdasr6Jl","type":"other","filename":"package","fileSize":6644,"extension":".json","mimeType":"text/plain"}% 
```

tomaszdurka replied:
👍

roosnic1 commented on docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+      context: .
+      target: app
+    ports:
+      - 4001
```
I suggest to use the port 6000.
We use 3000 for the editor, 4000 for the API and 5000 for the website. And now we could set the media server at 6000.

tomaszdurka replied:
This is dynamic port exposure - meaning it will be different port every time - do you suggest hardcoding this port? I prefer this way especially because of the conflicts in ports you might have. Still this requires usage of `docker-compose port` / `docker port` commands.

roosnic1 replied:
If you just do a docker-compose up then it will be port `4001` to access the media server. So even if we work with dynamic ports if we work with some standard ports like the ones i mentioned above i think it will make development easier. And if somebody pulls this image they know it's port 5000.

roosnic1 commented on docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+      NUM_CLUSTERS: 1
+      STORAGE_PATH: ./.media
+      TOKEN: super-secret
+    volumes:
```
I think it would make sense to add the `./.media` folder to the volumes. Otherwise all the images will be gone if somebody does a `docker-compose down`. In the `package.json` is already a script `clean:media` that will remove the folder.

roosnic1 commented on src/index.ts:
```diff
@@ -0,0 +1,45 @@
+  startMediaServer({
+    storageBackend: new LocalStorageBackend(storagePath),
+    imageBackend: new SharpImageBackend(),
+    maxUploadSize: 2048 * 2048 * 10, // 10
```
I suggest to make this also configurable by ENV var. The developers from Bajour are working on a PR to resize images above a certain file size.

tomaszdurka committed:
Adjust to review
- Add setup command for copying .env
- Streamline TOKEN values
- Adjust node12 source image
- Add local development readme section
- Add .media volume to docker-compose.yml
- Add MAX_UPLOAD_SIZE env variable

tomaszdurka committed:
Add github action

tomaszdurka committed:
Improve doc, add doc about production builds

tomaszdurka committed:
Temporarily build docker image on pull-requests events

tomaszdurka committed:
Use context instead of variable in build-push tags

tomaszdurka committed:
Fix target to app-production

tomaszdurka committed:
Revert image back to 12.20

tomaszdurka committed:
Rename workflow to proper

tomaszdurka committed:
Revert back to master pushes only

roosnic1 commented on docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+version: '3.7'
```
when i run `docker-compose up` i get the following error:
```
media_1  | yarn run v1.22.5
media_1  | $ npm-run-all --parallel watch:*
media_1  | /bin/sh: npm-run-all: not found
media_1  | error Command failed with exit code 127.
media_1  | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
karma-media-server_media_1 exited with code 127

```

roosnic1 commented on .github/workflows/docker-image-release.yml:
```diff
@@ -0,0 +1,30 @@
+on:
+  push:
+    branches:
+      - 'master'
```
On new repositories on Github the default branch is called `main`

roosnic1 reviewed (commented):
Looks really good just a few more comments

tomaszdurka committed:
chore(app): Improvements in documentation, docker-compose and port changes

tomaszdurka committed:
chore(app): Dockerfile improvements

roosnic1 reviewed (approved):
🎉
